caput/mpiarray.py

```diff
@@ -1036,6 +1036,177 @@ def _partition_io(self, skip=False, threshold=1.99):
+        # that the ufunc knows how to work with
+        outputs = kwargs.get("out", None)
+        if outputs:
+            out_args, _ = MPIArray._mpi_to_ndarray(outputs)
+            kwargs["out"] = tuple(out_args)
+        else:
+            outputs = (None,) * ufunc.nout
+
+        results = super().__array_ufunc__(ufunc, method, *args, **kwargs)
+
+        if results is NotImplemented:
+            return NotImplemented
+
+        # operation is performed in-place, so we can just return
+        if method == "at":
+            return
+
+        if ufunc.nout == 1:
+            results = (results,)
+
+        if "reduce" in method:
+            # reduction methods eliminate axes, so the distributed axes needs to be recalculated
+            if results[0].shape and len(results[0].shape) < dist_axis + 1:
+                dist_axis = len(results[0].shape) - 1
+
+        # Wrapping the results back into MPIArrays, distributed across the appropriate axis
+        ret = []
+
+        for result, output in zip(results, outputs):
+            if (
+                output is not None
+            ):  # case: results were placed in the array specified by `out`; return as is
+                ret.append(output)
+            else:
+                if (
+                    result.shape
+                ):  # case: the result is an array; convert back it into an MPIArray
+                    ret.append(MPIArray.wrap(result, axis=dist_axis))
+                else:  # case: result is a scalar; convert to 1-d vector, distributed across axis 0
+                    ret.append(MPIArray.wrap(np.reshape(result, (1, 1)), axis=0))
+
+        return ret[0] if len(ret) == 1 else tuple(ret)
+
+    # pylint: enable=inconsistent-return-statements
+
+    def __array_finalize__(self, obj):
+        """
+        Finalizes the creation of the MPIArray, when viewed.
+
+        In NumPy, ndarrays only go through the `__new__` when being instantiated.
+        For views and broadcast, they go through __array_finalize__.
+        https://numpy.org/doc/stable/user/basics.subclassing.html#the-role-of-array-finalize
+
+        Parameters
+        ----------
+        obj : MPIArray or None
+            The original MPIArray being viewed or broadcast. When in the middle of a constructor, obj is set to None.
+        """
+        # we are in the middle of a constructor, and the attributes
+        # will be set when we return to it
+        if obj is None:
+            return
+
+        # we are in a ufunc, rebuild the attributes from the original MPIArray
+        if isinstance(obj, MPIArray):
+            comm = getattr(obj, "comm", mpiutil.world)
+            axis = getattr(
+                obj, "axis", 0
+            )  # probably not a good default! How would we find this out?
+
+            # if the array has been reduced, re-calibrate the distr axes
+            if len(self.shape) < axis + 1:
+                axis = len(self.shape) - 1
```

Comment:
I don't think this clause is correct, as it doesn't seem to depend on which axis was the reduction axis.

I would check your test coverage here. I think something like this should fail with the clause as above.
```
dist_arr = MPIArray((3, 3, 3, 3), comm=MPI.COMM_WORLD, axis=1)
dist_arr.sum(axis=0)
```
I think the issue is that the arrays in the tests below aren't long enough to trigger this error.